rust/xaynet-sdk/src/client/reqwest.rs

```diff
@@ -20,81 +17,73 @@ use xaynet_core::{
     UpdateSeedDict,
 };
 
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 /// A client that communicates with the coordinator's API via HTTP(S).
-pub struct HttpApiClient {
+pub struct Client {
     /// HTTP client
-    client: Client,
+    client: reqwest::Client,
     /// Coordinator URL
-    address: String,
+    address: Arc<String>,
 }
 
-impl HttpApiClient {
+impl Client {
     /// Creates a new HTTP(S) client.
     ///
     /// Requires trusted server `certificates` if the `tls` feature is enabled.
-    pub fn new<S>(
-        address: S,
-        #[cfg(feature = "tls")] certificates: Vec<Certificate>,
-    ) -> Result<Self, HttpApiClientError>
+    pub fn new<S>(address: S, certificates: Option<Vec<Certificate>>) -> Result<Self, ClientError>
     where
         S: Into<String>,
     {
-        #[cfg(not(feature = "tls"))]
-        let client = ClientBuilder::new()
-            .build()
-            .map_err(HttpApiClientError::Http)?;
+        let address = Arc::new(address.into());
+        if certificates.is_none() {
```

Comment:
i agree that we should have either or, but not both mixed together.

that said, i'm more in favor of feature flags currently to keep the binary small when compiled without unwanted features. on the other hand, since this is now an example implementation of a client provided by us, we could simply say that we only provide such an example impl with tls and if anybody wants some without tls they are free to impl themselves. this way we could get rid of tls feature flags and options on the client side all together.

Reply:
On the long run I was thinking that the client could not implement any HTTP related logic and could just accept a reqwest client. The user builds the reqwest client they want, and pass it to `xaynet_sdk::Client`. `xaynet_sdk::Client` just calls the reqwest client to make the requests. That way we don't have to worry about the APIs around certificates for instance.

It would also make it quite easy to support other http clients like `surf` in a similar way: the user builds the http client themself, and pass it to our wrapper which uses it to make the requests.